### PR TITLE
Authentication completed

### PR DESCRIPTION
**Pull request by Patience250**
### What does this PR do?
Validating blogs and queries
### Description of tasks to be completed
- Install passport,passport-jwt
- Validating inputs
- Creating a token 
- Assign a token to the user
- Protect blog routes
### How to test manually
- Clone the repository then open the folder ```server``` 
- start server with command "npm run dev"
- Open postman and to POST on "localhost:5000/api/users/login" with email:test@gmail.com, password:testingapi@123

### CODE REVIEW

File: server/src/middlewares/privateRoutes.js [hunk 1]
```diff
@@ -0,0 +1,33 @@
+const jwt = require("jsonwebtoken")
+const User = require("../models/User")
+
+module.exports.auth = (req, res, next) => {
+    const token = req.header("auth-token")
+    if (!token) return res.status(401).send({ error: "Access Denied" })
+    try {
+        const verified = jwt.verify(token, "skdhkshdkjhskhkhkjshfjkfsfd")
```
Comment (niyodusengaclement) [inline server/src/middlewares/privateRoutes.js]: Use .env

File: server/src/utils/utils.js [hunk 2]
```diff
@@ -0,0 +1,26 @@
+const jsonwebtoken = require('jsonwebtoken');
+const fs = require('fs');
+const path = require('path');
+
+const pathToKey = path.join(__dirname, '..', 'id_rsa_priv.pem');
+const PRIV_KEY = fs.readFileSync(pathToKey, 'utf8');
+/**
+ * @param {*} user - The user object.  We need this to set the JWT `sub` payload property to the MongoDB user ID
```
Comment (niyodusengaclement) [inline server/src/utils/utils.js:8]: What is it for?

File: server/src/middlewares/privateRoutes.js [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+        next()
+    } else {
+        res.status(401).send({ error: "Access Denied" })
+    }
+}
+
+module.exports.hasToken = (passport) => {
+        return passport.authenticate('jwt', { session: false })
+    }
+    // module.exports.isAdmin = (req, res, next) => {
```
Comment (niyodusengaclement) [inline server/src/middlewares/privateRoutes.js]: No dead code.
Delete all commented lines

File: server/src/config/passport.js [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+//     done(null, user.id);
+// });
+
+// passport.deserializeUser(function(id, done) {
+//     User.findById(id, function(err, user) {
+//         done(err, user);
+//     });
+// });
+
+
```
Comment (niyodusengaclement) [inline server/src/config/passport.js]: Delete commented lines

File: node_mongo_tutorial/src/controllers/books.js [hunk 5]
```diff
@@ -0,0 +1,41 @@
+const Book = require("../models/Book")
+
+
+exports.findBooks = async(req, res) => {
+    const books = await Book.find()
+    res.send({ data: books })
+}
+
+exports.createBook = async(req, res) => {
+    const book = new Book(req.body)
```
Comment (niyodusengaclement) [inline node_mongo_tutorial/src/controllers/books.js:10]: You need Input Validation

File: server/.env [hunk 6]
```diff
@@ -0,0 +1,8 @@
+DB_STRING = "mongodb://localhost:27017/my-brand"
```
Comment (niyodusengaclement) [inline server/.env:1]: Do not ever push .env

File: server/src/controllers/users.js [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+        console.log(error)
+        res.status(400).send({ error: error.message })
+    }
+
+}
+
+// get 
+
+exports.getUsers = async(req, res) => {
+    const users = await User.find()
```
Comment (niyodusengaclement) [inline server/src/controllers/users.js]: Use try and catch in every function

**Review by niyodusengaclement — CHANGES_REQUESTED** [review]
Fix the feedback and check if they are no same errors that I have not mentioned

File: server/src/utils/utils.js (hunk 2 above)
Comment (Patience250) [reply server/src/utils/utils.js:8]: It is reading that file that includes private key, if i remove that PRIV KEY the returned in token string , it throws this error "error": "secret must be a string or bufferor a KeyObject"

**Pull request merged**